geopandas/sindex.py

```diff
@@ -1,25 +1,124 @@
+    Dynamically chooses a spatial indexing backend.
+    Required to comply with _compat.USE_PYGEOS.
+    The order of preference goes PyGeos > RTree > None.
     """
+    if compat.USE_PYGEOS:
+        return PyGEOSSTRTreeIndex
+    elif compat.HAS_RTREE:
+        return RTreeIndex
+    else:
+        warn("Cannot generate spatial index: Missing package `rtree` or `pygeos`.")
+        return None
+
+
+if compat.HAS_RTREE:
+
+    from rtree.index import Index as ToblerityRTreeIndex  # noqa
+    from rtree.core import RTreeError  # noqa
+
+    warn(
+        "`rtree` will be deprecated in GeoPandas 0.9; "
+        "please use `pygeos` instead: https://github.com/pygeos/pygeos"
+    )
+
+    class RTreeIndex(ToblerityRTreeIndex):
+        """
+        A simple wrapper around rtree's RTree Index
+        """
+
+        def __init__(self, geometry):
+            stream = (
+                (i, item.bounds, idx)
+                for i, (idx, item) in enumerate(geometry.iteritems())
+                if notnull(item) and not item.is_empty
+            )
+            try:
+                super().__init__(stream)
+            except RTreeError:
+                # What we really want here is an empty generator error, or
+                # for the bulk loader to log that the generator was empty
+                # and move on.
+                # See https://github.com/Toblerity/rtree/issues/20.
+                super().__init__()
+
+        @property
+        def size(self):
+            return len(self.leaves()[0][1])
+
+        @property
+        def is_empty(self):
+            if len(self.leaves()) > 1:
+                return False
+            return self.size < 1
+
+
+if compat.HAS_PYGEOS:
+
+    from pygeos import STRtree, box, Geometry  # noqa
+
+    class PyGEOSSTRTreeIndex(STRtree):
+        """
+        A simple wrapper around pygeos's STRTree
+        """
+
+        with_objects = namedtuple("with_objects", "object id")
+
+        def __init__(self, geometry):
+            # for compatibility with old RTree implementation, store ids/indexes
+            original_indexes = geometry.index
+            non_empty = geometry[~geometry.values.is_empty]
+            self.objects = self.ids = original_indexes[~geometry.values.is_empty]
+            super().__init__(non_empty.values.data)
+
+        def _process_query(self, geometry, op, objects):
+            if isinstance(geometry, (tuple, list)) and len(geometry) == 4:
+                # bounds, convert to geometry
+                # this is for RTree compatibility
+                indexes = super().query(box(*geometry), predicate=op)
+                if objects:
+                    objs = self.objects[indexes].values
+                    ids = self.ids[indexes]
+                    return [
+                        self.with_objects(id=id, object=obj)
+                        for id, obj in zip(ids, objs)
+                    ]
+                else:
+                    return indexes
```

Comment:
Is the `objects` parameter of `rtree` being used anywhere? I only see it used in tests, and it adds quite a bit of complexity to these wrappers.

I think it's important to have some type of way to choose if we want to work with iloc style indexing of label based indexing (i.e. have `sindex` return label based indexes), and that indeed needs to live in this wrapper, but there is something to be said for writing that interface from scratch instead of trying to conform to what `rtree` offers.

Reply:
> Is the objects parameter of rtree being used anywhere? I only see it used in tests, and it adds quite a bit of complexity to these wrappers.

I don't think so. I would say it is certainly not needed to mimic this part for the pygeos wrapper. But again, let's leave deprecation for later, since rtree is still the default for now.

Reply:
Will leave as is for now, but certainly something to consider for a future PR.